### PR TITLE
A fix allowing to convert numpy types (np.bool, np.int, np.float) to sympy Float

### LINKED ISSUE #13275: creating Float explicitly from numpy types can fail
#11666 improved various `S(<numpy thing>)` but `Float(<numpy thing>)` still gives errors like `TypeError: cannot create mpf from...`

Examples:
````
Float(np.int64(5))
Float(np.float128(5))
Float(np.float32(5))
````

`Float(np.float64(5))` is ok on my computer, possibly because `np.float64` is the same as native python double-precision `float`.

Probably be nice to fix this without duplicating lots of code from the #11666 fix but I haven't looked at it at all so I could be wrong.

### PR DESCRIPTION
**Pull request by Trave11er**
The code from #11666 was factored into a separate function (inside sympify.py) that can be called by both sympify and Float. Now can do

Float(np.int64(5))
Float(np.float128(5),12)
Float(np.float32(5.1200000),dps=5)

These previously gave an error but now evaluate to:

5
5.00000000000
5.1200

Corresponding tests have been added in sympy/core/tests/test_numbers.py
Fixes #13275

### CODE REVIEW

**Comment by Trave11er** [comment]
Have tried to all the failed tests tests locally and they do succeed. Not sure where to proceed from here - any guidance would be appreciated.

File: sympy/core/tests/test_numbers.py [hunk 1]
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
     assert Float('1.0', precision=Integer(15))._prec == 15
     assert type(Float('1.0', precision=Integer(15))._prec) == int
     assert sympify(srepr(Float('1.0', precision=15))) == Float('1.0', precision=15)
+
+def test_numpy():
+    from sympy.utilities.pytest import skip
+    import numpy as np
```
Comment (ylemkimon) [inline sympy/core/tests/test_numbers.py]: You have to do `np = import_module('numpy')`, otherwise it raises `ImportError`.

**Review by ylemkimon — CHANGES_REQUESTED** [review]
Other Travis test failures are caused by dependent modules.

File: sympy/core/tests/test_numbers.py (hunk 1 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Thank you - have changed the imports. Regarding Travis - does it meant I have to make sure local dependencies (and their versions) are the same as those used in the automatic tests?

Comment (ylemkimon) [reply sympy/core/tests/test_numbers.py]: @Trave11er I meant that it's their problem(s) and not caused by your commits.

File: sympy/core/sympify.py [hunk 2]
```diff
@@ -53,7 +55,7 @@ class CantSympify(object):
 # Support for basic numpy datatypes
 
 def convert_numpy_to_sympy(a):
-    import numpy as np
+    np=import_module('numpy')
```
Comment (jksuom) [inline sympy/core/sympify.py]: Add a space on either side of =.

Comment (Trave11er) [reply sympy/core/sympify.py]: Have done (staged for the next commit)

File: sympy/core/sympify.py [hunk 3]
```diff
@@ -50,6 +52,23 @@ class CantSympify(object):
     """
     pass
 
+# Support for basic numpy datatypes
+
```
Comment (cbm755) [inline sympy/core/sympify.py]: two blank lines after previous class/def and no blank line following this comment,

File: sympy/core/sympify.py [hunk 4]
```diff
@@ -50,6 +52,23 @@ class CantSympify(object):
...
+def convert_numpy_to_sympy(a):
+    np=import_module('numpy')
```
Comment (cbm755) [inline sympy/core/sympify.py]: Someone can call this with a non-numpy object (?).  I think we should assert np is non-None here, as a sanity check

File: sympy/core/tests/test_sympify.py [hunk 5]
```diff
@@ -557,7 +559,7 @@ def equal(x, y):
         return x == y and type(x) == type(y)
 
     if not np:
-        skip('numpy not installed.Abort numpy tests.')
+        skip('np not installed.Abort np tests.')
```
Comment (cbm755) [inline sympy/core/tests/test_sympify.py]: ?

File: sympy/core/tests/test_sympify.py [hunk 6]
```diff
@@ -2,7 +2,7 @@
     Function, I, S, N, sqrt, srepr, Rational, Tuple, Matrix, Interval, Add, Mul,
     Pow, Or, true, false, Abs, pi, Range)
 from sympy.abc import x, y
-from sympy.core.sympify import sympify, _sympify, SympifyError, kernS
+from sympy.core.sympify import sympify, _sympify, SympifyError, kernS, convert_numpy_to_sympy
```
Comment (cbm755) [inline sympy/core/tests/test_sympify.py]: ?

File: sympy/core/tests/test_numbers.py [hunk 7]
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
...
+def test_numpy():
+    from sympy.utilities.pytest import skip
+    np = import_module('numpy')
+    def equal(x, y):
+        return x == y and type(x) == type(y)
+
+    if not np:
+        skip('numpy not installed.Abort numpy tests.')
+
+    assert Float(np.bool_(1)) is S(True)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: I'm not sure we should require `Float(bool)` to work, so no reason for it to work with `np.bool_`.

File: sympy/core/tests/test_numbers.py [hunk 8]
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
...
     assert sympify(srepr(Float('1.0', precision=15))) == Float('1.0', precision=15)
+
+def test_numpy():
+    from sympy.utilities.pytest import skip
+    np = import_module('numpy')
+    def equal(x, y):
+        return x == y and type(x) == type(y)
+
+    if not np:
+        skip('numpy not installed.Abort numpy tests.')
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: move up before the `def equal` and fix grammar (space after period).

File: sympy/core/tests/test_numbers.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
 
 from mpmath import mpf
 import mpmath
-
+from sympy.external import import_module
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: move down to test function?

File: sympy/core/tests/test_numbers.py [hunk 10]
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
...
+
+    if not np:
+        skip('numpy not installed.Abort numpy tests.')
+
+    assert Float(np.bool_(1)) is S(True)
+    try:
+        assert equal(
+            Float(np.int_(12345678912345678)), S(12345678912345678.))
+        assert equal(
+            Float(np.intp(12345678912345678)), S(12345678912345678.))
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: A lot of these tests don't look right to me.  For example here, you can comparing the `S(<int>)` to `S(<float>)`.  They indeed could be the same but its not really what we're supposed to be testing.

Is this whole test a copy-paste with minor edits from sympify?  Maybe that is what we want but all those RHS with `.` are not right.

File: sympy/core/numbers.py [hunk 11]
```diff
@@ -1021,6 +1027,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
 
         if isinstance(num, float):
             _mpf_ = mlib.from_float(num, precision, rnd)
+        elif type(num).__module__ == 'numpy':
+            _mpf_ = mlib.from_float(num, precision, rnd)
```
Comment (cbm755) [inline sympy/core/numbers.py]: Eventually we probably want to do this.  But I'm not sure np floats work in all possible paths through the above code for dealing with all the special cases for precision and dps.  For example, does `Float(np...., '')` work?

**Review by cbm755 — CHANGES_REQUESTED** [review]
Good ideas here but I'm not sure the Float constructor bits are correct yet.

And test set needs more thought and polish,

File: sympy/core/tests/test_numbers.py (hunk 10 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: These tests are indeed a copy-paste from test_sympify. Could you please clarify your comment - my motivation for adding `.` on the RHS was to make sure that the types agree with those on the LHS which all belong to `S(<float>)`.

File: sympy/core/numbers.py (hunk 11 above)
Comment (Trave11er) [reply sympy/core/numbers.py]: Will remove if you insist but my suggestion is to keep it allows us to do things like
```
Float(np.float128(5),12)
Float(np.float32(5.1200000),dps=5)
```
Which is better than nothing. The example, you gave indeed fails, but so does `Float(<int>, '')` so it is a more general issue.

**Review by Trave11er — COMMENTED** [review]
Have incorporated all requested changes, except the code for 1) dealing with special cases of precision and dps 2) '.' on the RHS of assertions in test_numbers.py.

File: sympy/core/tests/test_numbers.py (hunk 10 above)
Comment (bmgxyz) [reply sympy/core/tests/test_numbers.py]: I agree with @Trave11er. The LHS is of type ```sympy.core.numbers.Float```, so for the RHS to have the same type, ```S()``` must take ```float``` as an input.

**Review by bmgxyz — COMMENTED** [review]
I can't quite speak to the implementation, but the tests seem good to me (especially if they've been copied from ```sympify.py```). I think the final fix for this issue - whatever form it may take - should include most or all of these tests.

File: sympy/core/sympify.py [hunk 12]
```diff
@@ -50,6 +50,25 @@ class CantSympify(object):
     """
     pass
 
+# Support for basic numpy datatypes
+
+def convert_numpy_to_sympy(a):
```
Comment (smichr) [inline sympy/core/sympify.py]: fwiw, I like that this is factored out of the sympify routine.

File: sympy/core/tests/test_numbers.py [hunk 13]
```diff
@@ -1749,3 +1748,43 @@ def test_Integer_precision():
...
+    assert equal(Float(np.uint8(123)), S(123.))
+    assert equal(Float(np.uint16(12345)), S(12345.))
+    assert equal(Float(np.uint32(1234567891)), S(1234567891.))
+    assert equal(
+        Float(np.uint64(1234567891234567891)), S(1234567891234567891.))
+    assert equal(Float(np.float32(1.123456)), Float(1.123456, precision=24))
+    assert equal(Float(np.float64(1.1234567891234)),
+                 Float(1.1234567891234, precision=53))
+    assert equal(Float(np.longdouble(1.123456789)),
+                 Float(1.123456789, precision=80))
```
Comment (asmeurer) [inline sympy/core/tests/test_numbers.py]: Is longdouble always precision 80? Could it be quad precision?

File: sympy/core/tests/test_numbers.py [hunk 14]
```diff
@@ -1749,3 +1748,43 @@ def test_Integer_precision():
...
+    assert equal(Float(np.complex64(1 + 2j)), S(1.0 + 2.0 * I))
+    assert equal(Float(np.complex128(1 + 2j)), S(1.0 + 2.0 * I))
+    assert equal(Float(np.longcomplex(1 + 2j)), S(1.0 + 2.0 * I))
```
Comment (asmeurer) [inline sympy/core/tests/test_numbers.py]: These should test the precision. Also, I don't know if there are discrepancies in any of these types across platforms, but if there are, we should guard against them.

Comment (asmeurer) [reply sympy/core/tests/test_numbers.py]: Actually, it seems kind of odd that Float should accept a complex number. `Float(1j)` currently doesn't work. I think these should actually be raising an exception, at least until https://github.com/sympy/sympy/pull/12192 is finished. You can test that something raises an exception with `raises`.

File: sympy/core/tests/test_numbers.py (hunk 13 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: I believe it is always 80 - the documentation states that longdouble is padded with zeros to either 96 or 128 bits.
Source: https://docs.scipy.org/doc/numpy-dev/user/basics.types.html

File: sympy/core/tests/test_numbers.py (hunk 14 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: To test the precision we would need Float instead of S on the RHS but as @asmeurer noticed Float does not currently accept complex numbers. It does accept them through numpy-to-sympy converting function though - so maybe add a raise in Float there whenever np.complex gets passed as an argument?

File: sympy/core/tests/test_numbers.py (hunk 13 above)
Comment (cbm755) [reply sympy/core/tests/test_numbers.py]: I think it is architecture dependent.

**Comment by cbm755** [comment]
For some reason there is no reply box, but this is a reply to...

> Could you please clarify your comment

I mean the stuff on the RHS of your tests would come from Rationals or strings rather than Python `float`.  That is, more like `Float(S(1)/3, precision=...)`.

My suggestion for tests looks something like this in pseudocode:
````
a = numpy.float32(1)/3
b = Float(a)
check that a.precision is same as b.precision
c = Float(S(1)/3, precision=b.precision)
check that relative difference between b and c is less than 2^{-precision}
````
Note, no hardcoded precisions here.  Then repeat that for .float64 and .float128.

For np.float128, also want:
````
a = numpy.float128(1)/3
b = Float(a, '')
check that b.precision is 53.
c = Float(S(1)/3, precision=b.precision)
check that b == c
````

And:
````
a = numpy.float128(1)/3
b = Float(a, precision=10)
check that b.precision is 10.
c = Float(S(1)/3, precision=b.precision)
check that b == c
````

Finally, I'm not convinced we need to test all the non-floating point numpy types here.  At least the non-floating-point-numpy should be in a different `test_` function.

**Comment by cbm755** [comment]
Ideally, Float(numpy.float128) would be loseless: it would have the same binary representation.  I'm not sure if `convert_numpy_to_sympy` does that yet or not.  But that could be a different issue/PR.  It would need its own (careful) test.

File: sympy/core/numbers.py (hunk 11 above)
Comment (cbm755) [reply sympy/core/numbers.py]: I didn't say remove.  I hope the other PR makes those things work.

I assume you mean `Float(<numpy.int>, '')`: please file bug if `Float<sympy int>, '')` isn't working.

File: sympy/core/tests/test_numbers.py (hunk 10 above)
Comment (cbm755) [reply sympy/core/tests/test_numbers.py]: As I tried to explain elsewhere, the RHS should be `Float('12345678912345678', something)`, for some value of `something`.  Bare Python `float`s should be avoided in these tests.

File: sympy/core/tests/test_numbers.py (hunk 14 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Added raise for np.complex64 and np.complex128.

File: sympy/core/tests/test_numbers.py (hunk 13 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: @cbm755 you are right - the long double comes from C and is general depends on architecture and compiler. Don't know an easy solution to test for different architectures so removed it along with `np.longcomplex` from test_numbers.py for now.

File: sympy/core/tests/test_numbers.py (hunk 10 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Have rewritten the tests as you outlined in you comment with pseudocode below (to which I also can't reply weirdly). Have changed numbers.py to be the same as committed by @bmgxyz and it passes all the new tests (also, e.g. `Float(np.float64(5),'')` works now).

File: sympy/core/numbers.py [hunk 15]
```diff
@@ -975,6 +977,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
 
         if dps is None and precision is None:
             dps = 15
+            if type(num).__module__ == 'numpy':
+                return convert_numpy_to_sympy(num)
```
Comment (cbm755) [inline sympy/core/numbers.py]: Perhaps this is dead code?  Can it be a numpy type when we get here?

File: sympy/core/sympify.py (hunk 3 above)
Comment (cbm755) [reply sympy/core/sympify.py]: But its not fixed?

File: sympy/core/tests/test_numbers.py [hunk 16]
```diff
@@ -18,7 +18,6 @@
 import mpmath
 
 
-
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: unrelated change

File: sympy/core/tests/test_numbers.py [hunk 17]
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
...
+
+def test_numpy_to_float(): 
+    def check_approximate_equality(numpyval, testnum):
+            fromnumpy = Float(numpyval)
+            fromS = Float(S(testnum), precision = fromnumpy._prec)
+            assert (np.finfo(numpyval).nmant == fromnumpy._prec)
+            assert(abs(fromnumpy-fromS)/fromS<2**(-fromnumpy._prec+1))
+
+    from sympy.external import import_module
+    np = import_module('numpy')
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: abort when np is None?

File: sympy/core/tests/test_numbers.py [hunk 18]
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
...
+
+    a1 = np.float32(testnum)
+    check_approximate_equality(a1, testnum)
+    a2 = np.float64(testnum)
+    check_approximate_equality(a2, testnum)
+    a3 = np.float128(testnum)
+    check_approximate_equality(a3, testnum)
+
+    fromnumpy = Float(a3)
+    fromS = Float(S(testnum), precision = fromnumpy._prec)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: `kwarg=a` without space for keyword arguments.  With space for assignment: `fromS = Float...` is correct.

File: sympy/core/tests/test_numbers.py [hunk 19]
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
...
+    assert(fromnumpy == fromS)
+
+    fromnumpy = Float(a3, precision = 10)
+    fromS = Float(S(testnum), precision = fromnumpy._prec)
+    assert(fromnumpy._prec == 10)
+    assert(fromnumpy == fromS)
+
+    raises(TypeError, lambda: Float(np.complex64(1+2j)))
+    raises(TypeError, lambda: Float(np.complex128(1+2j)))
+'''
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: ??

File: sympy/core/sympify.py (hunk 3 above)
Comment (Trave11er) [reply sympy/core/sympify.py]: Apologies - done now.

File: sympy/core/numbers.py (hunk 15 above)
Comment (Trave11er) [reply sympy/core/numbers.py]: Totally agree - removed.

File: sympy/core/tests/test_numbers.py (hunk 16 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Reverted.

File: sympy/core/tests/test_numbers.py (hunk 17 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Added an assertion.

File: sympy/core/tests/test_numbers.py (hunk 18 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Thank you for explaining - fixed.

File: sympy/core/tests/test_numbers.py (hunk 19 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: `Float(np.complex64(1j))` gives out an error as does `Float(1j)` acting on suggestion from @asmeurer above.

File: sympy/core/tests/test_numbers.py [hunk 20]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
     assert Float('1.0', precision=Integer(15))._prec == 15
     assert type(Float('1.0', precision=Integer(15))._prec) == int
     assert sympify(srepr(Float('1.0', precision=15))) == Float('1.0', precision=15)
+
+def test_numpy_to_float(): 
+    def check_approximate_equality(numpyval, testnum):
+            fromnumpy = Float(numpyval)
+            fromS = Float(S(testnum), precision = fromnumpy._prec)
+            assert (np.finfo(numpyval).nmant == fromnumpy._prec)
+            assert(abs(fromnumpy-fromS)/fromS<2**(-fromnumpy._prec+1))
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: `/fromS` also goes inside abs.  too much indenting here.  define this after you know you have np.

File: sympy/core/tests/test_numbers.py [hunk 21]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
...
+    a1 = np.float32(testnum)
+    check_approximate_equality(a1, testnum)
+    a2 = np.float64(testnum)
+    check_approximate_equality(a2, testnum)
+    a3 = np.float128(testnum)
+    check_approximate_equality(a3, testnum)
+
+    fromnumpy = Float(a3)
+    fromS = Float(S(testnum), precision=fromnumpy._prec)
+    assert(fromnumpy._prec == 63)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: maybe it would be safer just to assert its `> 53`?

File: sympy/core/tests/test_numbers.py [hunk 22]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
...
+    assert(fromnumpy == fromS)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: remove this (too strict I think)

File: sympy/core/tests/test_numbers.py [hunk 23]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
...
+
+    fromnumpy = Float(a3, precision=10)
+    fromS = Float(S(testnum), precision=fromnumpy._prec)
+    assert(fromnumpy._prec == 10)
+    assert(fromnumpy == fromS)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: here you could use `same_and_same_precision`

File: sympy/core/tests/test_numbers.py [hunk 24]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
...
+
+    raises(TypeError, lambda: Float(np.complex64(1+2j)))
+    raises(TypeError, lambda: Float(np.complex128(1+2j)))
+
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: extra blank line at end of file.

File: sympy/core/tests/test_numbers.py [hunk 25]
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
...
+        skip('numpy not installed. Abort numpy tests.')
+
+    testnum = 123456789
+
+    a1 = np.float32(testnum)
+    check_approximate_equality(a1, testnum)
+    a2 = np.float64(testnum)
+    check_approximate_equality(a2, testnum)
+    a3 = np.float128(testnum)
+    check_approximate_equality(a3, testnum)
```
Comment (cbm755) [inline sympy/core/tests/test_numbers.py]: I'd prefer a repeating decimal rather than an integer, to be sure the relative error check happens.

maybe:
````
    check_precision_and_relerr(np.float32(2)/3, S(2)/3)
    check_precision_and_relerr(np.float64(2)/3, S(2)/3)
    check_precision_and_relerr(np.float128(2)/3, S(2)/3)
````

**Review by cbm755 — CHANGES_REQUESTED** [review]

File: sympy/core/tests/test_numbers.py (hunk 20 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Yes, makes sense.

File: sympy/core/tests/test_numbers.py (hunk 25 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Agreed - thank you for clarifying.

File: sympy/core/tests/test_numbers.py (hunk 21 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Yes, but don't we want to be more strict? As I understand, `np.float128` should certainly have > 52 since `np.float64` has 52 bits in mantissa.

File: sympy/core/tests/test_numbers.py (hunk 23 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Totally agree.

File: sympy/core/tests/test_numbers.py (hunk 22 above)
Comment (Trave11er) [reply sympy/core/tests/test_numbers.py]: Agreed. Would it be worth adding instead `assert same_and_same_prec(y, Float(S(2)/3, precision=53))`

Comment (cbm755) [reply sympy/core/tests/test_numbers.py]: I don't think that is true.  y should have precision 63 (on my intel cpu).

File: sympy/core/tests/test_numbers.py (hunk 21 above)
Comment (cbm755) [reply sympy/core/tests/test_numbers.py]: We discussed this already: np.float128 is 63 bits on my computer, but I don't know what the minimum guaranteed is.  If you find a reference we can assert the minimum based on that.
